InvenTree/machine/registry.py

```diff
@@ -0,0 +1,155 @@
+import logging
+from typing import Dict, List, Union
+from uuid import UUID
+
+from machine.machine_type import BaseDriver, BaseMachineType
+
+logger = logging.getLogger('inventree')
+
+
+class MachinesRegistry:
+    def __init__(self) -> None:
+        """Initialize machine registry
+
+        Set up all needed references for internal and external states.
+        """
+
+        self.machine_types: Dict[str, BaseMachineType] = {}
+        self.drivers: Dict[str, BaseDriver] = {}
+        self.driver_instances: Dict[str, BaseDriver] = {}
+        self.machines: Dict[str, BaseMachineType] = {}
+
+        self.base_drivers: List[BaseDriver] = []
+        self.errors = []
+
+    def initialize(self):
+        print("INITIALIZE")  # TODO: remove debug statement
+        self.discover_machine_types()
+        self.discover_drivers()
+        self.load_machines()
+
+    def discover_machine_types(self):
+        import InvenTree.helpers
+
+        logger.debug("Collecting machine types")
+
+        machine_types: Dict[str, BaseMachineType] = {}
+        base_drivers: List[BaseDriver] = []
+
+        discovered_machine_types: List[BaseMachineType] = InvenTree.helpers.inheritors(BaseMachineType)
+        for machine_type in discovered_machine_types:
+            try:
+                machine_type.validate()
+            except NotImplementedError as error:
+                self.errors.append(error)
+                continue
+
+            machine_types[machine_type.SLUG] = machine_type
+            base_drivers.append(machine_type.base_driver)
```

Comment:
Where is the driver validated in this case?

Reply:
They don't need to, they are only the base drivers. Actually they won't validate without an error, because the only implement an abstract implementation. I only collected this list to subtract them from all inharitants.